### PR TITLE
Robust matched filter

### LINKED ISSUE #118: Implement Robust Matched Filter
For cases when the background clutter model and/or the target signal are not known with perfect fidelity, the standard matched filter can fail.  Robust matched filters apply some regularization to the problem to develop a more reliable result.  This task deals with this implementation.

References:

Manolakis, D., Lockwood, R., Cooley, T., & Jacobson, J. (2009, August). [Hyperspectral detection algorithms: Use covariances or subspaces?](https://dspace.mit.edu/bitstream/handle/1721.1/52735/Manolakis-2009-Hyperspectral%20detection%20algorithms%20Use%20covariances%20or%20subspaces.pdf?sequence=1&isAllowed=y). In Imaging Spectrometry XIV (Vol. 7457, p. 74570Q). International Society for Optics and Photonics.

### PR DESCRIPTION
**Pull request by jpolchlo**
## Overview

Target detection requires a target to search for, but it may not be entirely easy to acquire a proper target spectrum that is likely to be found in a given scene.  Available spectra may be lab sampled, and lack the characteristics of the target substance as they appear in the environment or as they are appear when sampled by a specific remote sensor.  This can be dealt with to some extent by using a robust matched filter, which allows the detected signature to deviate from the supplied signature by a small amount (a bound is placed on ‖s - s0‖, where s0 is the supplied target spectrum and s is the true target spectrum).

This PR provides an implementation of the robust matched filter.

Closes #118

## Demo

See the included notebook.

## Notes

When dealing with smaller samples in the generation of a background covariance model, it is advised to use a shrinkage estimator for the covariance.  There was an implementation in this library, but it hadn't been extensively used.  It is not a very numerically-robust estimator, and especially seems to fail for larger-dimensional inputs, which is unfortunately our use case.  The recommendation is to switch to `sklearn.covariance.LedoitWolf` to access a shrunk estimate.

## Checklist

- [ ] CI passes after rebase
- [X] ~~README.md updated if necessary to reflect the changes~~

### CODE REVIEW

**Comment by jamesmcclain** [comment]
:eyes:

**Comment by jamesmcclain** [comment]
I just have a few general comments.

1. If the MIF notebook changed in this PR, please describe how.
2. It was nice to see the maximum likelihood spectrum make an appearance in the RMF notebook.  It would also be interesting to test the default plastic spectrum from the library.  Just looking at the inferred spectrum from the gist (picture below), it looks like more iteration might have been needed to get a good spectrum (assuming that the method is capable of providing one).
3. Since the general scheme in Manolakis, et al. is pretty formal (but also with parameters) it might be interesting to empirically estimate some of those parameters using Pytorch (I am thinking of the loading factor and positive definite matrix that defines the ellipse, in particular).
4. The general idea of trading decreased variance for increased bias is interesting.  I think that the Pytorch-based optimization of the whitening operator can be thought of as expressing a similar idea -- the Pytorch code tries to the BCE loss (which is kind of like variance) and has the option to do so by squeezing some of that into a bias parameter.  It might be interesting to constrain that a bit more formally using some of the ideas from these papers.

image

**Review by jamesmcclain — CHANGES_REQUESTED** [review]
I think that a few more comments in the library code could help.  I tried to map some of the functions to equations in Manolakis, et al.: if those mappings are correct I think they should appear as comments in the code.

**Comment by jamesmcclain** [comment]
@jpolchlo I added the PRISMA notebooks to this PR branch, please feel free to remove the commit if you do not want those notebook in this branch.

**Comment by jamesmcclain** [comment]
@jpolchlo Should we try to get this merged?
